security/op-node-op-signer-integration.md

```diff
@@ -0,0 +1,145 @@
+
+The signing of block payload in op-node is a custom signing structure where the signing hash is: `[32]byte keccak256Hash([32]byte domain, [32]byte chainId, [32]byte keccak256(payloadBytes))` (see op-node/p2p/signer.go). This signing hash is signed with the private key, to produce a ECDSA signature of 65 bytes. 
+
+# Proposed Solution
+
+Above design can be applied to op-node to allow for remote signing of new block payloads in a sequencer. 
+
+For this, we need changes in the following modules: 
+
+1. op-node to import corresponding `opsigner` module from the op-service
+
+This module allows the user to choose from private-key / mnemonic / remote signer and create a single Signer from the factory. This involves importing `op-service/signer` and `op-service/crypto` into op-node.
+
+2. Modify `op-service/signer` to handle signing of block payload, in addition to tx signing. 
+
+Currently, `op-service` only has `SignTransactions` function which calls `eth_signTranasction`. To abstractify the signing method for different use cases, we might add different signing strategy types, or add new functions like `SignBlockPayload`. 
+
+1. Modify `op-signer` to serve different signing requests.
+
+Currently, `op-signer` only handles `eth_signTrasnaction`. We need to introduce a new rpc endpoint to handle signing of block payload
+
+The diagram illustrates what the result op-stack service looks like:
+
+```mermaid
+sequenceDiagram
+    participant PublicPeers as public peers
+    participant Sequencer as sequencer op-node
+    participant OpSigner as op-signer
+    participant GCP as GCP HSM (or any other service)
+
+    Sequencer ->> OpSigner: block data payload
+    OpSigner -->> Sequencer: mTLS verify
+    Sequencer -->> OpSigner: mTLS verify
+    OpSigner ->> GCP: Request Signing of payload
+    GCP ->> OpSigner: Return Signature
+    OpSigner ->> Sequencer: Return signature
+    Sequencer ->> PublicPeers: Propagate block w/ signature
+
+```
+
+## New RPC: eth_signBlockPayload
+
+new eth-domain RPC (or under some other namespace)
+
+```go
+func eth_signBlockPayload(signingHash [32]byte) (sig *[65]byte, err error) 
+// where signingHash is keccak256Hash([32]byte domain, [32]byte chainId, [32]byte keccak256(payloadBytes))
+```
+
+This new RPC will be added to:
+
+1. optimism/op-service/signer/client.go: new SignBlockPayload function which calls, the eth_signBlockPayload RPC. This method is called by the op-node’s p2p/signer.go
+2. optimism-infra/op-signer/service/service.go: new SignBlockPayload, which is the receiver counterpart of the RPC. 
+
+Most implementation of this RPC will be decoupling the signing logic from op-node, and moving code around different parts of the monorepo/op-signer. 
+
+### Considerations
+
+Above signing structure is a custom scheme, and does not align with other more official signing schemes like the ones introduced by EIP-712 or ERC-191, so we can’t use something like `eth_sign` or `eth_signTypedData`. 
```

Comment:
It seems desirable to try and align on a more standard signing scheme. Provided it doesn't add considerable complexity / additional engineering effort, I recommend we implement against a standard and preserve backwards compatibility for older versions.

Reply:
I agree on using standard signing scheme and adding backwards compatibility to existing scheme when possible. I did look at other schemes like `eth_sign` and `eth_signTypedData`, but they require more payload and parsing as they contain ethereum-related information necessary for signing. 

I think this rpc is specific to the op-node's p2p signing behavior, and will most likely just be handled by the op-signer(or may beother optimism-related clients in the future), rather than being used in general clients like geth, etc. It'd be best to minimize the payload size to keep the signature latency as minimal possible, and that would require optimism-specific rpc. 

Rather, what do you think about changing the rpc name from `eth_signBlockPayload` to something like `optimism_signBlockPayload` to remove the impression that this operation is related to ethereum when it's not?

Reply:
+1 one introducing it into the `optimism` RPC namespace, or a new `opsigner` namespace, rather than the ethereum namespace.

Signing with `eth_signTypedData` over typed data would make sense, but that API also has a few flaws (more complicated schema decoding and communication, on every signing request). We can look into unifying that in a future upgrade, if desired. Perhaps `optimism_signBlockPayload` can become a front-end, wrapping `eth_signTypedData` and applying the schema, type checks and logging needed for block signing.

Reply:
I updated the design doc and the implementation to use the new `opsigner` namespace: https://github.com/ethereum-optimism/infra/pull/59/commits/dda416c364ab20f97c945845c901106c3df4d771. And left unifying the signing scheme to a possible future work in the docs

Reply:
+1 to not adding to the `eth` namespace. It doesn't look like neither https://github.com/ethereum-optimism/infra/pull/59 in the signer client nor https://github.com/ethereum-optimism/optimism/pull/12325 docs were updated to the new name.

Reply:
@trianglesphere Oh you're right, the code were updated but not the descriptions. I just updated the description for those PRs to the new `opsigner_signBlockPayload`